File: dotnet/Mcma.Worker/Builders/WorkerBuilder.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Mcma.Worker.Builders
{
    public class WorkerBuilder
    {
        private List<IOperationHandlerBuilder> OperationHandlerBuilders { get; } = new List<IOperationHandlerBuilder>();

        public WorkerBuilder HandleOperation<T>(string operationName, Action<OperationHandlerBuilder<T>> configureOperation)
        {
            var opHandlerBuilder = new OperationHandlerBuilder<T>(operationName);
            OperationHandlerBuilders.Add(opHandlerBuilder);
            configureOperation(opHandlerBuilder);
            return this;
        }

        public IWorker Build() => new Worker(OperationHandlerBuilders.SelectMany(r => r.Build()));
    }
}

File: nodejs/base/worker/lib/builders/worker-builder.js
const { OperationHandlerBuilder } = require("./operation-handler-builder");
const { Worker } = require("../worker");

class WorkerBuilder {
    constructor() {
        const operationHandlerBuilders = {};

        this.handleOperation = (operationName, configureOperation) => {
            if (!configureOperation) {
                if (typeof operationName === "function" && operationName.name !== "Function") {
                    const operation = operationName;
                    configureOperation = x => x.handle(operation);
                    operationName = operation.name;
                }
            }
            if (!operationName || typeof operationName !== "string" || operationName.length === 0) {
                throw new Error("operationName must be a non-empty string.");
            }
            if (!configureOperation || typeof configureOperation !== "function") {
                throw new Error("configureOperation must be a function");
            }
            if (Object.keys(operationHandlerBuilders).find(k => k.toLowerCase() === operationName.toLowerCase())) {
                throw new Error("Operation with name '" + operationName + "' is already being handled.");
            }
            const opHandlerBuilder = new OperationHandlerBuilder(operationName);
            operationHandlerBuilders[operationName] = opHandlerBuilder;
            configureOperation(opHandlerBuilder);
            return this;
        };

        this.build = () => new Worker(Object.values(operationHandlerBuilders).reduce((flattened, opBuilder) => flattened.concat(opBuilder.build()), []));
    }
}

module.exports = {
    WorkerBuilder
};
